File: projects/punk0/code/shift_right.go
package main

import "fmt"

// shift to the input list to the right,
// adding 0 at the end:
//   [1,2,3,4]
// returns:
//   [0,1,2,3]
func shift_right(x []int) []int {
	r := []int{}

	// start with a zero
	r = append(r, 0)

	// copy everything except the last
	// element
	for i := 0; i < len(x)-1; i++ {
		v := x[i]
		r = append(r, v)
	}

	return r
}

func main() {
	fmt.Printf("%v\n", shift_right([]int{1, 1, 2, 3, 3, 4, 1, 2, 7, 9}))
}

File: projects/punk0/code/shift_right.c
#include <stdio.h>
#include <stdlib.h>
typedef struct list {
  int len;
  int *data;
} list;

// shift to the input list to the right,
// adding 0 at the end:
//   [1,2,3,4]
// returns:
//   [0,1,2,3]
list shift_right(list x) {
  // start with len=0 and allocate space
  // for x.len elements, 4 bytes each
  list r = {0, malloc(x.len * 4)};
  // start with a zero
  r.data[r.len++] = 0;
  // copy everything except the last
  // element
  for (int i = 0; i < x.len-1; i++) {
    int v = x.data[i];
    r.data[r.len++] = v;
  }

  return r;
}

int main(void) {
  list x = {
      .len = 10,
      .data = malloc(x.len * 4),
  };

  int n = 0;
  x.data[n++] = 1;
  x.data[n++] = 1;
  x.data[n++] = 2;
  x.data[n++] = 3;
  x.data[n++] = 3;
  x.data[n++] = 4;
  x.data[n++] = 1;
  x.data[n++] = 2;
  x.data[n++] = 7;
  x.data[n++] = 9;

  list r = shift_right(x);
  printf("[");
  for (int i = 0; i < r.len; i++) {
    printf("%d", r.data[i]);
    if (i != r.len - 1) {
      printf(" ");
    }
  }
  printf("]\n");
}
